File: haskell/src/Graph/BreadthFirstSearch.hs
module Graph.BreadthFirstSearch (bfs) where

import Data.List

import Core.Queue
import Graph.DirectedGraph

-- breadth first traversal of a directed graph
bfs :: (Eq a) => Node a -> [a]
bfs start = bfsHelper [start] []

doesNotContain :: (Eq a) => [a] -> a -> Bool
doesNotContain l x = not $ elem x l

isNew :: (Eq a) => Queue (Node a) -> [Node a] -> Node a -> Bool
isNew q visited n = and [doesNotContain q n, doesNotContain visited n]

bfsHelper :: (Eq a) => Queue (Node a) -> [Node a] -> [a]
bfsHelper [] _ = []
bfsHelper q visited = let curr = dequeue q -- get the next node
                          currNode = fst curr
                          currQueue = snd curr
                          currValue = value currNode
                          currNeighbors = neighbors currNode -- get the neighbors of the current node
                          newNeighbors = filter (isNew q visited) currNeighbors -- filter neighbors to ones that aren't enqueued or visited
                          newQueue = currQueue ++ newNeighbors -- updated queue for next iteration
                          newVisited = currNode : visited
                      in
                          currValue : (bfsHelper newQueue newVisited) -- recurse on the rest of the queue with updated visited list

File: go/src/graphs/breadth_first_search.go
package graphs

import "../core"

func (d *DirectedNode) BreadthFirstSearch() []int {
    q := new(core.Queue)
    q.Enqueue(d)

    var visited []*DirectedNode

    for n, _ := q.Dequeue(); q.Size() != 0; n, _ = q.Dequeue() {
        curr := n.(*DirectedNode)

        for _, neighbor := range curr.neighbors {
            if !(sliceContains(visited, neighbor) || q.Contains(neighbor, compareNode)) {
                q.Enqueue(neighbor)
            }
        }

        visited = append(visited, curr)
    }

    // extract the actual values
    var out []int

    for _, n := range visited {
      out = append(out, n.value)
    }

    return out
}
